File: src/meta_schedule/strategy/postproc.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include "./postproc.h"  // NOLINT(build/include)

#include "../analysis.h"

namespace tvm {
namespace meta_schedule {

/********** Constructor **********/

Postproc::Postproc(String name, FProc proc) {
  ObjectPtr<PostprocNode> n = make_object<PostprocNode>();
  n->name = std::move(name);
  n->proc_ = std::move(proc);
  data_ = std::move(n);
}

/********** Postproc **********/

bool PostprocNode::Apply(const Schedule& sch, Sampler* sampler) { return proc_(sch, sampler); }

Array<Postproc> PostprocDefaults() { return {RewriteParallel(), RewriteVectorize()}; }

/********** Built-in Post-Processors **********/

Postproc RewriteParallel() {
  auto f_proc = [](Schedule self, void* _sampler) -> bool {
    Array<Array<tir::StmtSRef>> to_parallel = CollectAnnotatedLoops(self->sch, "lazy_parallel");
    for (const Array<tir::StmtSRef>& group : to_parallel) {
      for (const tir::StmtSRef& loop_sref : group) {
        const auto* loop = loop_sref->GetStmt<tir::LoopNode>();
        CHECK(loop) << "TypeError: Expects LoopNode, but gets: " << loop_sref->GetTypeKey();
        ObjectPtr<tir::LoopNode> new_loop = make_object<tir::LoopNode>(*loop);
        new_loop->annotations.clear();
        self->sch->Replace(loop_sref, tir::Loop(new_loop));
      }
      tir::StmtSRef fused = group[0];
      for (int i = 1, n = group.size(); i < n; ++i) {
        fused = self->sch->fuse(fused, group[i]);
      }
      self->sch->parallel(fused);
    }
    return true;
  };
  return Postproc("rewrite_parallel", f_proc);
}

Postproc RewriteVectorize() {
  auto f_proc = [](Schedule self, void* _sampler) -> bool {
    Array<Array<tir::StmtSRef>> to_vectorize = CollectAnnotatedLoops(self->sch, "lazy_vectorize");
    for (const Array<tir::StmtSRef>& group : to_vectorize) {
      for (const tir::StmtSRef& loop_sref : group) {
        const auto* loop = loop_sref->GetStmt<tir::LoopNode>();
        CHECK(loop) << "TypeError: Expects LoopNode, but gets: " << loop_sref->GetTypeKey();
        ObjectPtr<tir::LoopNode> new_loop = make_object<tir::LoopNode>(*loop);
        new_loop->annotations.clear();
        self->sch->Replace(loop_sref, tir::Loop(new_loop));
      }
      tir::StmtSRef fused = group[0];
      for (int i = 1, n = group.size(); i < n; ++i) {
        fused = self->sch->fuse(fused, group[i]);
      }
      self->sch->vectorize(fused);
    }
    return true;
  };
  return Postproc("rewrite_vectorize", f_proc);
}

/********** FFI **********/

struct Internal {
  /*!
   * \brief FFI function for PostProcNode::Apply
   * \sa PostProcNode::Apply
   */
  static bool Apply(Postproc self, Schedule sch, Optional<Integer> seed) {
    Sampler seeded;
    if (seed.defined()) {
      seeded.Seed(seed.value());
    }
    return self->Apply(sch, &seeded);
  }
};

TVM_REGISTER_NODE_TYPE(PostprocNode);
TVM_REGISTER_GLOBAL("meta_schedule.postproc.Apply").set_body_typed(Internal::Apply);
TVM_REGISTER_GLOBAL("meta_schedule.postproc.RewriteParallel").set_body_typed(RewriteParallel);
TVM_REGISTER_GLOBAL("meta_schedule.postproc.RewriteVectorize").set_body_typed(RewriteVectorize);

}  // namespace meta_schedule
}  // namespace tvm

File: python/tvm/meta_schedule/postproc.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""Postprocessors are applied at the final stage before a schedule is sent to be measured"""

from typing import List, Optional

from tvm._ffi import register_object
from tvm.runtime import Object
from tvm.tir import Schedule, TensorIntrin

from . import _ffi_api_postproc
from .search import SearchTask


@register_object("meta_schedule.Postproc")
class Postproc(Object):
    """A post processor, used for the search strategy for postprocess the schedule it gets"""

    name: str

    def apply(
        self,
        task: SearchTask,
        sch: Schedule,
        seed: Optional[int] = None,
    ) -> bool:
        """Postprocess the schedule.

        Parameters
        ----------
        sch : Schedule
            The schedule to be mutated
        seed : Optional[int]
            The random seed

        Returns
        -------
        result : bool
            If the post-processing succeeds
        """
        return _ffi_api_postproc.Apply(self, task, sch, seed)  # pylint: disable=no-member


def rewrite_tensorize(tensor_intrins: List[TensorIntrin]) -> Postproc:
    """Creates a postprocessor that matches the region that is marked as auto tensorized

    Returns
    ----------
    postproc: Postproc
        The postprocessor created
    """
    return _ffi_api_postproc.RewriteTensorize(tensor_intrins)  # pylint: disable=no-member


def rewrite_tensorize_tensor_core(
    compute_intrin: str,
    load_intrin_A: str,
    load_intrin_B: str,
    store_intrin: str,
    init_intrin: str,
) -> Postproc:
    """Creates a postprocessor that matches the region that is marked as auto tensorized

    Parameters
    ----------
    compute_intrin: str
        The tensor intrinsinc for doing computation
    load_intrin_A: str
        The corresponding data load intrinsic for compute_intrin
    load_intrin_B: str
        The corresponding data load intrinsic for compute_intrin
    store_intrin: str
        The corresponing data store instrinsic for compute_intrin
    init_intrin: str
        The intrinsic for tensor core computation

    Returns
    ----------
    postproc: Postproc
        The postprocessor created
    """
    return _ffi_api_postproc.RewriteTensorizeTensorCore(
        compute_intrin, load_intrin_A, load_intrin_B, store_intrin, init_intrin
    )  # pylint: disable=no-member


def rewrite_cooperative_fetch() -> Postproc:
    """Creates a postprocessor rewrites "lazy_cooperative_fetch" with the actual threadIdx

    Returns
    ----------
    postproc: Postproc
        The postprocessor created
    """
    return _ffi_api_postproc.RewriteCooperativeFetch()  # pylint: disable=no-member


def rewrite_cooperative_fetch_tensor_core() -> Postproc:
    """Creates a postprocessor rewrites "lazy_cooperative_fetch" with the threadIdx==32

    Returns
    ----------
    postproc: Postproc
        The postprocessor created
    """
    return _ffi_api_postproc.RewriteCooperativeFetchTensorCore()  # pylint: disable=no-member


def rewrite_unbound_blocks() -> Postproc:
    """Creates a postprocessor that finds each block that is not bound to thread axes,
    and bind them to `blockIdx.x` and `threadIdx.x`

    Returns
    ----------
    postproc: Postproc
        The postprocessor created
    """
    return _ffi_api_postproc.RewriteUnboundBlocks()  # pylint: disable=no-member


def rewrite_parallel_vectorize_unroll() -> Postproc:
    """Creates a postprocessor that applies parallelization, vectorization and auto unrolling,
    according to the annotation of each block

    Returns
    ----------
    postproc: Postproc
        The postprocessor created
    """
    return _ffi_api_postproc.RewriteParallelizeVectorizeUnroll()  # pylint: disable=no-member


def disallow_dynamic_loops() -> Postproc:
    """Create a postprocessor that checks if all loops are static

    Returns
    ----------
    postproc: Postproc
        The postprocessor created
    """
    return _ffi_api_postproc.DisallowDynamicLoops()  # pylint: disable=no-member


def verify_gpu_code() -> Postproc:
    """Creates a postprocessor that do block/vthread/thread binding for cuda

    Returns
    ----------
    postproc: Postproc
        The postprocessor created
    """
    return _ffi_api_postproc.VerifyGPUCode()  # pylint: disable=no-member


def rewrite_reduction_block() -> Postproc:
    """Creates a postprocessor that decomposes ReduceStep

    Returns
    ----------
    postproc: Postproc
        The postprocessor created
    """
    return _ffi_api_postproc.RewriteReductionBlock()  # pylint: disable=no-member


def rewrite_reduction_block_tensor_core() -> Postproc:
    """Creates a postprocessor that decomposes ReduceStep

    Returns
    ----------
    postproc: Postproc
        The postprocessor created
    """
    return _ffi_api_postproc.RewriteReductionBlockTensorCore()  # pylint: disable=no-member


def rewrite_layout() -> Postproc:
    """Creates a postprocessor that rewrites layout

    Returns
    ----------
    postproc: Postproc
        The postprocessor created
    """
    return _ffi_api_postproc.RewriteLayout()  # pylint: disable=no-member
